src/api/procedures/addInstruction.ts

```diff
@@ -98,128 +118,198 @@ export async function prepareAddInstruction(
     });
   }
 
-  if (!legs.length) {
+  const addAndAffirmInstructions: [
+    u64,
+    SettlementType,
+    Moment | null,
+    Moment | null,
+    {
+      from: PortfolioId;
+      to: PortfolioId;
+      asset: Ticker;
+      amount: Balance;
+    }[],
+    PortfolioId[]
+  ][] = [];
+  const addInstructions: [
+    u64,
+    SettlementType,
+    Moment | null,
+    Moment | null,
+    {
+      from: PortfolioId;
+      to: PortfolioId;
+      asset: Ticker;
+      amount: Balance;
+    }[]
+  ][] = [];
+
+  const noLegs: number[] = [];
+  const noFutureBlock: number[] = [];
+  const noFutureDate: number[] = [];
+
+  await P.each(instructions, async ({ legs, endBlock, tradeDate, valueDate }, i) => {
```

Comment:
COMPLEXITY: Let's try first refactoring this whole loop into a `getTxArgsAndErrors` function defined outside the prepare method that returns the necessary arrays as properties of an object